File: miataru/KnownDevice.m
//
//  KnownDevice.m
//  miataru
//
//  Created by Daniel Kirstenpfad on 07.09.13.
//  Copyright (c) 2013 Miataru. All rights reserved.
//

#import "KnownDevice.h"

//#define RGB(r, g, b) [UIColor colorWithRed:r/255.0 green:g/255.0 blue:b/255.0 alpha:1]

@implementation KnownDevice

+ (id) DeviceWithName:(NSString*)inName DeviceID:(NSString*)inDeviceID;
{
    KnownDevice *device = [[KnownDevice alloc] init];

    device.DeviceName = inName;
    device.DeviceID = inDeviceID;
    device.DeviceColor = [UIColor redColor];
   
    return device;
}

-(void)setUpdateTime:(NSDate*)NewUpdateDateTime
{
    self.LastUpdate = NewUpdateDateTime;
}

-(void)encodeWithCoder:(NSCoder *)aCoder
{
    [aCoder encodeObject:self.DeviceName forKey:@"DeviceName"];
    [aCoder encodeObject:self.DeviceID forKey:@"DeviceID"];
    [aCoder encodeObject:self.LastUpdate forKey:@"LastUpdate"];
    [aCoder encodeObject:self.DeviceColor forKey:@"DeviceColor"];
    [aCoder encodeBool:self.DeviceIsInGroup forKey:@"DeviceIsInGroup"];
    
    if (CLLocationCoordinate2DIsValid(self.LastKnownLocation))
    {
        NSString *hash = [GeoHash hashForLatitude: self.LastKnownLocation.latitude
                                longitude: self.LastKnownLocation.longitude
                                   length: 13];
        // geohash encode the location...
        [aCoder encodeObject:hash forKey:@"LastKnownLocation"];
    }
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
    if (! (self = [super init])) {
        return nil;
    }

    self.DeviceID = [aDecoder decodeObjectForKey:@"DeviceID"];
    self.DeviceName = [aDecoder decodeObjectForKey:@"DeviceName"];
    self.LastUpdate = [aDecoder decodeObjectForKey:@"LastUpdate"];
    self.DeviceColor = [aDecoder decodeObjectForKey:@"DeviceColor"];
    self.DeviceIsInGroup = [aDecoder decodeBoolForKey:@"DeviceIsInGroup"];
    
    NSString *hash = [aDecoder decodeObjectForKey:@"LastKnownLocation"];
    
    if (hash)
    {
        GHArea *area = [GeoHash areaForHash:hash];
        self.LastKnownLocation = CLLocationCoordinate2DMake(area.latitude.min.doubleValue, area.longitude.min.doubleValue);
    }
    return self;
}


@end

File: liquid-miataru/miataru/miataru/SettingsManagers/App Settings/KnownDevices/KnownDevice.swift
//
//  KnownDevice.swift
//  miataru
//
//  Created by Daniel Kirstenpfad on 20.06.25.
//  Copyright © 2025 Miataru. All rights reserved.
//

import Foundation
import Combine
import UIKit
import MapKit

@objc(KnownDevice)
class KnownDevice: NSObject, ObservableObject, NSCoding, NSSecureCoding, Identifiable {
    @Published @objc var DeviceName: String
    @Published @objc var DeviceID: String
    @Published @objc var DeviceIsInGroup: Bool = false
    @Published @objc var KnownDevicesTablePosition: Int = 0
    @Published @objc var DeviceColor: UIColor?
    
    var id: String { DeviceID }
    
    init(name: String, deviceID: String, color: UIColor? = nil) {
        self.DeviceName = name
        self.DeviceID = deviceID
        self.DeviceColor = color
    }
    
    required init?(coder aDecoder: NSCoder) {
        self.DeviceName = aDecoder.decodeObject(forKey: "DeviceName") as? String ?? ""
        self.DeviceID = aDecoder.decodeObject(forKey: "DeviceID") as? String ?? ""
        self.DeviceIsInGroup = aDecoder.decodeBool(forKey: "DeviceIsInGroup")
        self.KnownDevicesTablePosition = aDecoder.decodeInteger(forKey: "KnownDevicesTablePosition")
        let decodedColor = aDecoder.decodeObject(forKey: "DeviceColor") as? UIColor
        self.DeviceColor = decodedColor
        print(aDecoder.decodeObject(forKey: "DeviceName") as? String ?? "")
        print(aDecoder.decodeObject(forKey: "DeviceID") as? String ?? "")
        print(aDecoder.decodeBool(forKey: "DeviceIsInGroup"))
        print(aDecoder.decodeInteger(forKey: "KnownDevicesTablePosition"))
        print(String(describing: decodedColor))
    }

    func encode(with aCoder: NSCoder) {
        aCoder.encode(DeviceName, forKey: "DeviceName")
        aCoder.encode(DeviceID, forKey: "DeviceID")
        aCoder.encode(DeviceIsInGroup, forKey: "DeviceIsInGroup")
        aCoder.encode(KnownDevicesTablePosition, forKey: "KnownDevicesTablePosition")
        aCoder.encode(DeviceColor, forKey: "DeviceColor")
        //print("Speichern ist temporär deaktiviert - muss repariert werden!!!")

    }

    static var supportsSecureCoding: Bool {
        return true
    }

    static func DeviceWithName(_ inName: String, deviceID inDeviceID: String) -> KnownDevice {
        return KnownDevice(name: inName, deviceID: inDeviceID)
    }

    //func setUpdateTime(_ NewUpdateDateTime: Date) {
    //    self.LastUpdate = NewUpdateDateTime
    //}
}
